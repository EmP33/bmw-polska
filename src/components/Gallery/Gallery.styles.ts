import styled from "styled-components";

export const Container = styled.div`
  width: 100%;
  display: grid;
  grid-template-columns: repeat(12, 1fr);
  grid-template-rows: repeat(2, 18vw);
  row-gap: 0.625rem;
  column-gap: 0.625rem;
  margin-bottom: 50rem; // TEMP
  overflow: hidden;

  @media only screen and (max-width: 1200px) {
    margin-top: 5rem;
  }
  @media only screen and (max-width: 900px) {
    row-gap: 0.5rem;
    column-gap: 0.5rem;
  }
  @media only screen and (max-width: 600px) {
    row-gap: 0.25rem;
    column-gap: 0.25rem;
    grid-template-rows: repeat(4, 40vw);
    grid-template-columns: repeat(6, 1fr);
  }

  img {
    width: 100%;
    height: 100%;
    object-fit: cover;
    display: block;

    &:nth-child(1) {
      grid-column: 1/7;
      grid-row: 1/-1;

      @media only screen and (max-width: 600px) {
        grid-column: 1/-1;
        grid-row: 1/3;
      }
    }
    &:nth-child(2) {
      grid-column: 7/9;
      @media only screen and (max-width: 600px) {
        grid-column: 1/3;
        grid-row: 3/4;
      }
    }
    &:nth-child(3) {
      grid-column: 7/10;
      grid-row: 2;
      @media only screen and (max-width: 600px) {
        grid-column: 1/4;
        grid-row: 4/-1;
      }
    }
    &:nth-child(4) {
      grid-column: 9/-1;
      @media only screen and (max-width: 600px) {
        grid-column: 3/-1;
        grid-row: 3/4;
      }
    }
    &:nth-child(5) {
      grid-column: 10/-1;
      grid-row: 2;
      @media only screen and (max-width: 600px) {
        grid-column: 4/-1;
        grid-row: 4/-1;
      }
    }
  }
`;
