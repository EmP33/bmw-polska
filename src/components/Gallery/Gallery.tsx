import React from "react";
// Hooks
import useMediaQuery from "../../hooks/useMediaQuery";
// Styles
import { Container } from "./Gallery.styles";
// Images
import insideFront4k from "../../assets/images/inside-front@2x.webp";
import insideBack4k from "../../assets/images/inside-back@2x.webp";
import foto34k from "../../assets/images/foto3_1@2x.webp";
import sliderPhoto from "../../assets/images/slider1.webp";
import foto24k from "../../assets/images/foto2_2@2x.webp";
import insideFront from "../../assets/images/inside-front.webp";
import insideBack from "../../assets/images/inside-back.webp";
import foto3 from "../../assets/images/foto3_1.webp";
import foto2 from "../../assets/images/foto2_2.webp";

const Gallery = () => {
  const matches = useMediaQuery("(max-width: 900px)");
  return (
    <Container>
      <img
        data-aos="zoom-in"
        data-aos-offset="700"
        src={matches ? insideFront : insideFront4k}
        alt="BMW iX inside front"
      />
      <img
        data-aos="zoom-in"
        data-aos-offset="1000"
        src={matches ? foto3 : foto34k}
        alt="BMW iX back"
      />
      <img
        data-aos="zoom-in"
        data-aos-offset="1000"
        src={matches ? insideBack : insideBack4k}
        alt="BMW iX inside back"
      />
      <img
        data-aos="zoom-in"
        data-aos-offset="1000"
        src={sliderPhoto}
        alt="BMW iX front"
      />
      <img
        data-aos="zoom-in"
        data-aos-offset="1000"
        src={matches ? foto2 : foto24k}
        alt="BMW iX back"
      />
    </Container>
  );
};

export default Gallery;
