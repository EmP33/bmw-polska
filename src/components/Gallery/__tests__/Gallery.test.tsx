import Gallery from "../Gallery";
import { render, screen, fireEvent } from "@testing-library/react";
import ReactShallowRenderer from "react-test-renderer/shallow";
import { act } from "react-dom/test-utils";

describe("<Gallery />", () => {
  const renderProps = (matches?: boolean = false) => {
    Object.defineProperty(window, "matchMedia", {
      writable: true,
      value: jest.fn().mockImplementation((query) => ({
        matches: matches, // matches
        media: query,
        onchange: null,
        addListener: jest.fn(), // Deprecated
        removeListener: jest.fn(), // Deprecated
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
      })),
    });
  };
  it("should render properly", () => {
    renderProps();
    const renderer = new ReactShallowRenderer();
    renderer.render(<Gallery />);
    expect(renderer.getRenderOutput()).toMatchSnapshot();
  });
  it("should show lower quality images on smaller screen size", () => {
    renderProps(true);
    render(<Gallery />);
    const imageElements: HTMLImageElement[] = screen.getAllByRole("img");

    for (let img of imageElements) {
      expect(img.src).not.toContain("@2x");
    }
  });

  it("should show higher quality images on screen size bigger than 900px", () => {
    renderProps();
    render(<Gallery />);
    const imageElements: HTMLImageElement[] = screen.getAllByRole("img");

    expect(imageElements[0].src).toContain("@2x");
    expect(imageElements[1].src).toContain("@2x");
    expect(imageElements[2].src).toContain("@2x");
    expect(imageElements[4].src).toContain("@2x");
  });

  it("fullscreen photo should not be in the document without clicking on any of imageElements", () => {
    renderProps();
    render(<Gallery />);

    expect(
      screen.queryByAltText("Full screen bmw photo")
    ).not.toBeInTheDocument();
  });

  it("should show full screen photo when any of imageElements are clicked", async () => {
    renderProps();
    render(<Gallery />);
    const imageElements: HTMLImageElement[] = screen.getAllByRole("img");

    fireEvent.click(imageElements[0]);

    expect(
      await screen.findByAltText("Full screen bmw photo")
    ).toBeInTheDocument();
  });
});
