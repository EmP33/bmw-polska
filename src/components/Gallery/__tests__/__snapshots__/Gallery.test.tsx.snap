// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Gallery /> should render properly 1`] = `
<styled.div
  id="gallery"
>
  <img
    alt="BMW iX inside front"
    data-aos="zoom-in"
    src="inside-front@2x.webp"
  />
  <img
    alt="BMW iX back"
    data-aos="zoom-in"
    src="foto3_1@2x.webp"
  />
  <img
    alt="BMW iX inside back"
    data-aos="zoom-in"
    src="inside-back@2x.webp"
  />
  <img
    alt="BMW iX front"
    data-aos="zoom-in"
    src="slider1.webp"
  />
  <img
    alt="BMW iX back"
    data-aos="zoom-in"
    src="foto2_2@2x.webp"
  />
</styled.div>
`;
